esp-hal-common/src/i2c.rs

```diff
@@ -435,114 +432,267 @@ pub trait Instance {
-            }
-        }
-
-        // The different chips have highly very different timing configurations, so
-        // we're setting these up separately (this might introduce some overhead,
-        // but improves readability)
-        cfg_if::cfg_if! {
-            if #[cfg(any(esp32c2, esp32c3))] {
-                let scl_wait_high = if bus_freq.raw() <= 50000 { 0 } else { half_cycle / 8 };
-                let scl_high = half_cycle - scl_wait_high;
-                let sda_hold = half_cycle / 4;
-                let sda_sample = scl_high / 2;
-            } else if #[cfg(esp32s3)] {
-                let scl_high = if bus_freq.raw() <= 50000 { half_cycle } else { half_cycle / 5 * 4 + 4 };
-                let scl_wait_high = half_cycle - scl_high;
-                let sda_hold = half_cycle / 2;
-                let sda_sample = half_cycle / 2;
-            } else if #[cfg(esp32s2)] {
-                let scl_high = half_cycle / 2 + 2;
-                let scl_wait_high = half_cycle - scl_high;
-                let sda_hold = half_cycle / 2;
-                let sda_sample = half_cycle / 2 - 1;
-            } else {
-                // ESP32 is default (as it is the simplest case and does not even have
-                // the wait_high field)
-                let scl_high = half_cycle;
-                let sda_hold = half_cycle / 2;
-                let sda_sample = scl_high / 2;
-                let tout = half_cycle * 20;
-            }
-        }
+        let source_clk = source_clk.raw();
+        let bus_freq = bus_freq.raw();
 
+        let half_cycle: u32 = source_clk / bus_freq / 2;
         let scl_low = half_cycle;
-        let setup = half_cycle - 1;
-        let hold = half_cycle - 1;
-
-        #[cfg(not(esp32))]
-        let scl_low = scl_low as u16 - 1;
-
-        #[cfg(esp32)]
-        let scl_low = scl_low as u16;
-
-        let scl_high = scl_high as u16;
-
-        #[cfg(any(esp32c2, esp32c3, esp32s3))]
-        let scl_wait_high = scl_wait_high as u8;
+        let scl_high = half_cycle;
+        let sda_hold = half_cycle / 2;
+        let sda_sample = scl_high / 2;
+        let setup = half_cycle;
+        let hold = half_cycle;
+        let tout = half_cycle * 20; // default we set the timeout value to 10 bus cycles.
+
+        // SCL period. According to the TRM, we should always subtract 1 to SCL low
+        // period
+        let scl_low = scl_low - 1;
+        // Still according to the TRM, if filter is not enbled, we have to subtract 7,
+        // if SCL filter is enabled, we have to subtract:
+        //   8 if SCL filter is between 0 and 2 (included)
+        //   6 + SCL threshold if SCL filter is between 3 and 7 (included)
+        // to SCL high period
+        let mut scl_high = scl_high;
+        // In the "worst" case, we will subtract 13, make sure the result will still be
+        // correct
+
+        // NOTE since we always set the filter threshold to 7 we don't need conditional code here
```

Comment:
Mind changing "NOTE" to "FIXME" or something like that so we don't forget about it?